tests/conftest.py

```diff
@@ -283,17 +283,35 @@ def init_kwargs(request):
 
 
 @pytest.fixture(scope="function")
-def task(request, init_kwargs):
+def task(request, generate_task):
     """Gets a task instance."""
-    # set default values used for the initialization of the task.
-    init_args = {
-        "new_task_name": "",
-    }
+    new_task_name = _get_value_from_pytest_markers("new_task_name", request)
+    new_task_name = "" if new_task_name is None else new_task_name
+    return generate_task(task_name=new_task_name)
 
-    # iterate through markers and update arguments
+
+@pytest.fixture(scope="function")
+def generate_task(init_kwargs):
+    """Gets a factory function which can be used to generate new tasks."""
+    tasks_created = []
+
+    def _create_task(task_name=""):
+        task = nidaqmx.Task(new_task_name=task_name, **init_kwargs)
+        tasks_created.append(task)
+        return task
+
+    yield _create_task
+
+    # destroying the tasks created.
+    for task in tasks_created:
+        del task
```

Comment:
`del task` does not clean up the task. It only removes the binding for the `task` variable. This allows the task to be garbage collected, but it doesn't call the task's `close` method.
https://stackoverflow.com/questions/21053380/what-does-del-do-exactly

Note that https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#factories-as-fixtures calls `record.destroy()`, not `del record`.

However, `nidaqmx.Task` is a context manager, so this is a good use case for `ExitStack`:
```
with contextlib.ExitStack() as stack:
    def _create_task(task_name=""):
        return stack.enter_context(nidaqmx.Task(new_task_name=task_name, **init_kwargs))

    yield _create_task
```

**ALSO**

I'm not 100% convinced that it makes sense for this fixture to close the tasks. It would be simpler to leave that up to the callers (the `task` fixture and the few tests that need multiple tasks).

Reply:
This makes many tests fail with resource reservation errors.

Here are the errors from `pytest -v -k library`:
```
=============================================== short test summary info ===============================================
FAILED tests/component/test_write_properties.py::test___ao_current_task___get_string_list_property___returns_default_value[library_init_kwargs-aoTester] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/component/test_write_properties.py::test___ao_current_task__read_property___out_of_order___throws_daqerror[library_init_kwargs-aoTester] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_channels.py::TestChannels::test_co_channel[library_init_kwargs] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_exceptions.py::TestReadExceptions::test_timeout[library_init_kwargs] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_exceptions.py::TestReadExceptions::test_timeout_raw[library_init_kwargs] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_exceptions.py::TestReadExceptions::test_timeout_stream[library_init_kwargs] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestAnalogReadWrite::test_1_chan_1_samp[library_init_kwargs-3357959520953164899] - nidaqmx.errors.DaqWriteError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestAnalogReadWrite::test_n_chan_1_samp[library_init_kwargs-7853889532935124913] - nidaqmx.errors.DaqWriteError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestAnalogReadWrite::test_1_chan_n_samp[library_init_kwargs-4023432960782616046] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestAnalogReadWrite::test_n_chan_n_samp[library_init_kwargs-5540431315400418725] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestCounterReadWrite::test_count_edges_n_samp[library_init_kwargs-2980927874250788733] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestCounterReadWrite::test_pulse_freq_1_samp[library_init_kwargs-536888581506171183] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestCounterReadWrite::test_pulse_time_1_samp[library_init_kwargs-2136044199193667323] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestCounterReadWrite::test_pulse_ticks_1_samp[library_init_kwargs-7316891088165754464] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_1_chan_1_samp[library_init_kwargs-2794647051481591722-False] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_n_chan_1_samp[library_init_kwargs-6685494260096705145-output_enables0] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_n_chan_1_samp[library_init_kwargs-418285595378318854-output_enables1] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_n_chan_1_samp[library_init_kwargs-1238952956118557739-output_enables2] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_n_chan_1_samp[library_init_kwargs-2996837575829402075-output_enables3] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_1_chan_n_samp[library_init_kwargs-3043546904430488102-True] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_1_chan_n_samp[library_init_kwargs-2402937852528272093-False] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_n_chan_n_samp[library_init_kwargs-1683508228318656168-output_enables0] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_n_chan_n_samp[library_init_kwargs-6005071669319512282-output_enables1] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_n_chan_n_samp[library_init_kwargs-5624730316066202063-output_enables2] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_power_n_chan_n_samp[library_init_kwargs-1282216028392626114-output_enables3] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_read_write.py::TestPowerRead::test_mixed_chans_with_power[library_init_kwargs-9150725256568852309] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_analog_readers_writers.py::TestAnalogSingleChannelReaderWriter::test_one_sample[library_init_kwargs-1111254608853596258] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_analog_readers_writers.py::TestAnalogSingleChannelReaderWriter::test_many_sample[library_init_kwargs-5788731447215696140] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_analog_readers_writers.py::TestAnalogMultiChannelReaderWriter::test_one_sample[library_init_kwargs-7953990193493899618] - nidaqmx.errors.DaqWriteError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_analog_readers_writers.py::TestAnalogMultiChannelReaderWriter::test_many_sample[library_init_kwargs-6671643187784784661] - nidaqmx.errors.DaqWriteError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_counter_readers_writers.py::TestCounterReaderWriter::test_one_sample_uint32[library_init_kwargs-8135833188404745696] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_counter_readers_writers.py::TestCounterReaderWriter::test_multi_sample_uint32[library_init_kwargs-7830989384349696662] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_counter_readers_writers.py::TestCounterReaderWriter::test_one_sample_double[library_init_kwargs-6559599307050901102] - TypeError: 'Task' object is not callable
FAILED tests/legacy/test_stream_counter_readers_writers.py::TestCounterReaderWriter::test_multi_sample_double[library_init_kwargs-151054651979677578] - TypeError: 'Task' object is not callable
FAILED tests/legacy/test_stream_counter_readers_writers.py::TestCounterReaderWriter::test_one_sample_pulse_freq[library_init_kwargs-8341747965871546395] - TypeError: 'Task' object is not callable
FAILED tests/legacy/test_stream_counter_readers_writers.py::TestCounterReaderWriter::test_many_sample_pulse_freq[library_init_kwargs-4927803445952422204] - TypeError: 'Task' object is not callable
FAILED tests/legacy/test_stream_counter_readers_writers.py::TestCounterReaderWriter::test_one_sample_pulse_time[library_init_kwargs-784278945663897055] - TypeError: 'Task' object is not callable
FAILED tests/legacy/test_stream_counter_readers_writers.py::TestCounterReaderWriter::test_many_sample_pulse_time[library_init_kwargs-7477734035849746566] - TypeError: 'Task' object is not callable
FAILED tests/legacy/test_stream_counter_readers_writers.py::TestCounterReaderWriter::test_pulse_ticks_1_samp[library_init_kwargs-7596930336252949074] - TypeError: 'Task' object is not callable
FAILED tests/legacy/test_stream_counter_readers_writers.py::TestCounterReaderWriter::test_many_sample_pulse_ticks[library_init_kwargs-8889138780187751155] - TypeError: 'Task' object is not callable
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalSingleChannelReaderWriter::test_one_sample_port_uint16[library_init_kwargs-8070415402439250076] - nidaqmx.errors.DaqWriteError: Resource requested by this task has already been reserved by a different task.
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalSingleChannelReaderWriter::test_one_sample_port_uint32[library_init_kwargs-8497690205121978685] - nidaqmx.errors.DaqError: Resource requested by this task has already been reserved by a different task.
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalSingleChannelReaderWriter::test_many_sample_port_byte[library_init_kwargs-6607542416393089424] - nidaqmx.errors.DaqError: Resource requested by this task has already been reserved by a different task.
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalSingleChannelReaderWriter::test_many_sample_port_uint16[library_init_kwargs-8026877848656619394] - nidaqmx.errors.DaqError: Resource requested by this task has already been reserved by a different task.
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalMultiChannelReaderWriter::test_one_sample_one_line[library_init_kwargs-7068008452396278559] - nidaqmx.errors.DaqWriteError: Requested operation could not be performed, because the specified digital lines are e...
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalMultiChannelReaderWriter::test_one_sample_multi_line[library_init_kwargs-5640336099470418848] - nidaqmx.errors.DaqWriteError: Requested operation could not be performed, because the specified digital lines are e...
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalMultiChannelReaderWriter::test_one_sample_port_byte[library_init_kwargs-1464408375075823996] - nidaqmx.errors.DaqWriteError: Resource requested by this task has already been reserved by a different task.
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalMultiChannelReaderWriter::test_one_sample_port_uint16[library_init_kwargs-8082927952666580527] - nidaqmx.errors.DaqWriteError: Resource requested by this task has already been reserved by a different task.
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalMultiChannelReaderWriter::test_one_sample_port_uint32[library_init_kwargs-8056528193220385272] - nidaqmx.errors.DaqWriteError: Requested operation could not be performed, because the specified digital lines are e...
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalMultiChannelReaderWriter::test_many_sample_port_byte[library_init_kwargs-6049499019857682415] - nidaqmx.errors.DaqError: Resource requested by this task has already been reserved by a different task.
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalMultiChannelReaderWriter::test_many_sample_port_uint16[library_init_kwargs-1777743550799775187] - nidaqmx.errors.DaqError: Resource requested by this task has already been reserved by a different task.
FAILED tests/legacy/test_stream_digital_readers_writers.py::TestDigitalMultiChannelReaderWriter::test_many_sample_port_uint32[library_init_kwargs-5944050764703248129] - nidaqmx.errors.DaqError: Requested operation could not be performed, because the specified digital lines are either...
FAILED tests/legacy/test_stream_power_readers.py::TestPowerSingleChannelReader::test_power_1_chan_1_samp[library_init_kwargs-5677390728333224334-True] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerSingleChannelReader::test_power_1_chan_1_samp[library_init_kwargs-2812121736889157114-False] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerSingleChannelReader::test_power_1_chan_n_samp[library_init_kwargs-7681833076340284326-True] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerSingleChannelReader::test_power_1_chan_n_samp[library_init_kwargs-1776314849886653299-False] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerMultiChannelReader::test_power_n_chan_1_samp[library_init_kwargs-117224208181729175-output_enables0] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerMultiChannelReader::test_power_n_chan_1_samp[library_init_kwargs-2555248456974594430-output_enables1] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerMultiChannelReader::test_power_n_chan_1_samp[library_init_kwargs-2675678955043521680-output_enables2] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerMultiChannelReader::test_power_n_chan_1_samp[library_init_kwargs-3855007922252046835-output_enables3] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerMultiChannelReader::test_power_n_chan_n_samp[library_init_kwargs-1820402650310020898-output_enables0] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerMultiChannelReader::test_power_n_chan_n_samp[library_init_kwargs-536898653223400666-output_enables1] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
rgs-4636582271656235629-output_enables2] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerMultiChannelReader::test_power_n_chan_n_samp[library_init_kwargs-1679795331665009605-output_enables3] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerBinaryReader::test_power_1_chan_n_samp_binary[library_init_kwargs-6621489274339721819-True] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerBinaryReader::test_power_1_chan_n_samp_binary[library_init_kwargs-5974265500468199459-False] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerBinaryReader::test_power_n_chan_many_sample_binary[library_init_kwargs-766391235328514436-output_enables0] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerBinaryReader::test_power_n_chan_many_sample_binary[library_init_kwargs-3582288927432859046-output_enables1] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerBinaryReader::test_power_n_chan_many_sample_binary[library_init_kwargs-1237365205369514009-output_enables2] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_stream_power_readers.py::TestPowerBinaryReader::test_power_n_chan_many_sample_binary[library_init_kwargs-6472206274048436649-output_enables3] - nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_write_exceptions.py::TestWriteExceptions::test_overwrite[library_init_kwargs] - nidaqmx.errors.DaqWriteError: The specified resource is reserved. The operation could not be completed as specified.
FAILED tests/legacy/test_write_exceptions.py::TestWriteExceptions::test_overwrite_during_prime[library_init_kwargs] - AssertionError: assert -50103 == <DAQmxErrors.NO_MORE_SPACE: -200293>
============================ 72 failed, 146 passed, 345 deselected, 126 warnings in 13.42s ============================
```

ExitStack works as expected. I suppose it's better for the fixture to clean up, so that the tests are simpler. However, don't use this fixture for testing task init/close code; use the init_kwargs fixture instead.

Reply:
Thanks for the suggestion, i have updated the logic and all library interpreter tests are passing now.
I have also handled the tests that use task.close() differently using the `init_kwargs` in [test_resource_warnings.py](https://github.com/ni/nidaqmx-python/pull/323/files#diff-63ef856901b890b0e9d65b2c4ed1bd3f4b7401952b8265c09a1104ad72359c13:~:text=def%20test_task_double_close(,called%2C%20for%20the))